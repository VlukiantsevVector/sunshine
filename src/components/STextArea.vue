<template>
  <s-base-input
    :style="baseStyle"
    :inactive="inactive"
    :hasFocus="hasFocus"
    :isEmpty="isEmpty"
    :label="$attrs.label"
    :readonly="$attrs.readonly"
  >

    <textarea
      class="s-input__input s-input__textarea"
      :disabled="inactive"
      v-bind="$attrs"
      v-on="listeners"
      v-model="internalValue"
      @focus="hasFocus = true"
      @blur="hasFocus = false">
    </textarea>

  </s-base-input>
</template>
<script>

import Vue from 'vue';
import SBaseInput from './SBaseInput.vue';

export default Vue.extend({
  name: 's-text-area',
  inheritAttrs: false,

  components: {
    SBaseInput,
  },

  props: {
    value: {
      type: String,
      default: '',
    },
    width: {
      type: String,
      default: '10em',
    },
    height: {
      type: String,
      default: '10em',
    },
    maxHeight: {
      type: String,
      default: '10em',
    },
    inactive: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      internalValue: this.value,
      hasFocus: false,
    };
  },

  watch: {
    value(val) {
      this.internalValue = val;
    },
    internalValue(val) {
      if (val !== this.value) {
        this.$emit('input', val);
      }
    },
  },

  computed: {
    isEmpty() {
      return !this.internalValue;
    },
    listeners() {
      const
        {
          input, ...listeners
        } = this.$listeners;
      return listeners;
    },
    baseStyle() {
      return {
        height: this.height,
        width: this.width,
        alignItems: 'baseline',
      };
    },
  },
});

</script>
