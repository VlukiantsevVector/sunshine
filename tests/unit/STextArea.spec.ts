import { mount } from '@vue/test-utils';
import STextArea from '@/components/STextArea.vue';

describe('STextArea', () => {
  it('renders empty text field', () => {
    const wrapper = mount(STextArea, {
      propsData: {
        label: 'Text Area',
        value: '',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders text field with text', () => {
    const wrapper = mount(STextArea, {
      propsData: {
        label: 'Text Area',
        value: 'foobar',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('doesnt change value when read only', () => {
    const wrapper = mount(STextArea, {
      propsData: {
        label: 'Text Area',
        value: '',
        readonly: true,
      },
    });
    const textarea = wrapper.find('textarea');
    textarea.setValue('omg');
    textarea.trigger('change');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('doesnt change value when inactive', () => {
    const wrapper = mount(STextArea, {
      propsData: {
        label: 'Text Area',
        value: '',
        inactive: true,
      },
    });
    const textarea = wrapper.find('textarea');
    textarea.setValue('omg');
    textarea.trigger('change');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('emits entered value', () => {
    const wrapper = mount(STextArea, {
      propsData: {
        label: 'Text Area',
        value: '',
      },
    });
    const textarea = wrapper.find('textarea');
    textarea.setValue('omg');
    textarea.trigger('change');

    expect(wrapper.emitted('input')[0]).toEqual(['omg']);
  });
});
