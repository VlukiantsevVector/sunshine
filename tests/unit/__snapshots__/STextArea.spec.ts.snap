// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`STextArea doesnt change value when inactive 1`] = `
<div class="s-text-area s-input s-input--inactive"><span class="s-input__label">Text Area</span>
  <div class="s-input__textarea__wrapper s-input__textarea__wrapper--with-label"><textarea disabled="disabled" placeholder="" class="s-input__input s-input__textarea" style="height: auto;"></textarea></div>
  <!---->
</div>
`;

exports[`STextArea doesnt change value when read only 1`] = `
<div class="s-text-area s-input"><span class="s-input__label s-input__label--aside">Text Area</span>
  <div class="s-input__textarea__wrapper s-input__textarea__wrapper--with-label"><textarea readonly="readonly" placeholder="" class="s-input__input s-input__textarea" style="height: auto;"></textarea></div>
  <!---->
</div>
`;

exports[`STextArea renders empty text field 1`] = `
<div class="s-text-area s-input"><span class="s-input__label">Text Area</span>
  <div class="s-input__textarea__wrapper s-input__textarea__wrapper--with-label"><textarea rows="10" placeholder="" class="s-input__input s-input__textarea"></textarea></div>
  <!---->
</div>
`;

exports[`STextArea renders text field with text 1`] = `
<div class="s-text-area s-input"><span class="s-input__label s-input__label--aside">Text Area</span>
  <div class="s-input__textarea__wrapper s-input__textarea__wrapper--with-label"><textarea placeholder="" class="s-input__input s-input__textarea" style="height: auto;"></textarea></div>
  <!---->
</div>
`;
